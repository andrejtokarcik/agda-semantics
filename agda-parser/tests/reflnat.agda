module reflnat where


data ℕ : Set where
  Z : ℕ 
  S : ℕ -> ℕ 


_+_ : ℕ -> ℕ ->  ℕ
n + Z   = n
n + S m = S (n + m)

infixr 10 _+_

infixr 20 _*_
_*_ : ℕ -> ℕ ->  ℕ
n * Z   = Z
n * S m = n * m + n


data Bool : Set where
  tt : Bool
  ff : Bool

data ⊤ : Set where
  true : ⊤ 

data ⊥ : Set where


Atom : Bool -> Set
Atom tt = ⊤ 
Atom ff = ⊥ 

_==Bool_ : ℕ -> ℕ -> Bool
Z     ==Bool  Z     = tt
(S n) ==Bool  (S m) = n ==Bool m
_     ==Bool  _     = ff
-- Z     ==Bool (S _)  = ff
-- (S _) ==Bool Z      = ff


_==_ : ℕ -> ℕ -> Set
n == m = Atom ( n ==Bool m)


Refl : Set
Refl = (n : ℕ) -> n == n


refl : Refl
refl Z     = true
refl (S n) = refl n






