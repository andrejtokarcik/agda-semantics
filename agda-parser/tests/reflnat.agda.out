<k>
    .K
</k>
<mgu>
    subst(.KList)
</mgu>
<ctxstack>
    .List
</ctxstack>
<sig>
    Fun ( "Atom" ) |-> ( #symVariable(145) : Data ( "Bool" ) ) -> Set ( 0 )
    Fun ( "Refl" ) |-> Set ( 0 )
    Fun ( "_*_" ) |-> ( #symVariable(56) : Data ( "8469" ) ) -> ( #symVariable
      (59) : Data ( "8469" ) ) -> Data ( "8469" )
    Fun ( "_+_" ) |-> ( #symVariable(62) : Data ( "8469" ) ) -> ( #symVariable
      (165) : Data ( "8469" ) ) -> Data ( "8469" )
    Fun ( "_==Bool_" ) |-> ( #symVariable(139) : Data ( "8469" ) ) -> ( 
      #symVariable(142) : Data ( "8469" ) ) -> Data ( "Bool" )
    Fun ( "_==_" ) |-> ( #symVariable(154) : Data ( "8469" ) ) -> ( 
      #symVariable(157) : Data ( "8469" ) ) -> Set ( 0 )
    Fun ( "refl" ) |-> Fun ( "Refl" )
    Data ( "8469" ) |-> Set ( 0 )
    Data ( "8868" ) |-> Set ( 0 )
    Data ( "8869" ) |-> Set ( 0 )
    Data ( "Bool" ) |-> Set ( 0 )
    Con ( "S" ) |-> ( #symVariable(165) : Data ( "8469" ) ) -> Data ( "8469" )
    Con ( "Z" ) |-> Data ( "8469" )
    Con ( "ff" ) |-> Data ( "Bool" )
    Con ( "true" ) |-> Data ( "8868" )
    Con ( "tt" ) |-> Data ( "Bool" )
</sig>
<con>
    Data ( "8469" ) |-> l ( ListItem(Con ( "Z" )) ListItem(Con ( "S" )) )
    Data ( "8868" ) |-> l ( ListItem(Con ( "true" )) )
    Data ( "8869" ) |-> l ( .List )
    Data ( "Bool" ) |-> l ( ListItem(Con ( "tt" )) ListItem(Con ( "ff" )) )
</con>
<ctx>
    .Map
</ctx>
<meta>
    .Map
</meta>
<ncat>
    Name ( "8469" ) |-> Data ( "8469" )
    Name ( "8868" ) |-> Data ( "8868" )
    Name ( "8869" ) |-> Data ( "8869" )
    Name ( "Atom" ) |-> Fun ( "Atom" )
    Name ( "Bool" ) |-> Data ( "Bool" )
    Name ( "Refl" ) |-> Fun ( "Refl" )
    Name ( "S" ) |-> Con ( "S" )
    Name ( "Z" ) |-> Con ( "Z" )
    Name ( "_*_" ) |-> Fun ( "_*_" )
    Name ( "_+_" ) |-> Fun ( "_+_" )
    Name ( "_==Bool_" ) |-> Fun ( "_==Bool_" )
    Name ( "_==_" ) |-> Fun ( "_==_" )
    Name ( "ff" ) |-> Con ( "ff" )
    Name ( "refl" ) |-> Fun ( "refl" )
    Name ( "true" ) |-> Con ( "true" )
    Name ( "tt" ) |-> Con ( "tt" )
</ncat>
<fun>
    Fun ( "Atom" ) |-> FunClauses ( ListItem((#symFunClauses(72) Con ( "tt" ))
       = Data ( "8868" )) ListItem((#symFunClauses(76) Con ( "ff" )) = Data ( 
      "8869" )) )
    Fun ( "Refl" ) |-> ( Var ( "n" ) : Data ( "8469" ) ) -> ((Fun ( "_==_" ) 
      Var ( "n" )) Var ( "n" ))
    Fun ( "_*_" ) |-> FunClauses ( ListItem(((#symFunClauses(44) Var ( "n" )) 
      Con ( "Z" )) = Con ( "Z" )) ListItem(((#symFunClauses(68) Var ( "n" )) (
      Con ( "S" ) Var ( "m" ))) = ((Fun ( "_+_" ) ((Fun ( "_*_" ) Var ( "n" ))
       Var ( "m" ))) Var ( "n" ))) )
    Fun ( "_+_" ) |-> FunClauses ( ListItem(((#symFunClauses(15) Var ( "n" )) 
      Con ( "Z" )) = Var ( "n" )) ListItem(((#symFunClauses(36) Var ( "n" )) (
      Con ( "S" ) Var ( "m" ))) = (Con ( "S" ) ((Fun ( "_+_" ) Var ( "n" )) 
      Var ( "m" )))) )
    Fun ( "_==Bool_" ) |-> FunClauses ( ListItem(((#symFunClauses(83) Con ( 
      "Z" )) Con ( "Z" )) = Con ( "tt" )) ListItem(((#symFunClauses(104) (Con 
      ( "S" ) Var ( "n" ))) (Con ( "S" ) Var ( "m" ))) = ((Fun ( "_==Bool_" ) 
      Var ( "n" )) Var ( "m" ))) ListItem(((#symFunClauses(115) #symExpr(105))
       #symExpr(110)) = Con ( "ff" )) )
    Fun ( "_==_" ) |-> \ ( Var ( "n" ) : #symExpr(133) ) -> \ ( Var ( "m" ) : 
      #symExpr(136) ) -> (Fun ( "Atom" ) ((Fun ( "_==Bool_" ) Var ( "n" )) Var
       ( "m" )))
    Fun ( "refl" ) |-> FunClauses ( ListItem((#symFunClauses(163) Con ( "Z" ))
       = Con ( "true" )) ListItem((#symFunClauses(174) (Con ( "S" ) Var ( "n" 
      ))) = (Fun ( "refl" ) Var ( "n" ))) )
</fun>
